### ISSUE TITLE
Check metadata (geometry_name, crs) in merge/concat

### ISSUE THREAD

**Issue by jorisvandenbossche** [geopandas/geopandas#326]
Follow-up issue for https://github.com/geopandas/geopandas/pull/322

When concatting or merging multiple GeoDataFrames, we could:
- check if the `crs` is the same. If not, we can either raise, or convert on the fly to the crs of the first object. 
- check if the `_geometry_column_name` is equal. If not, raise?

**Comment by knaaptime** [comment]
I'm getting an odd error where pandas refuses to concatenate two geodataframes with the same CRS. I have two dataframes, both in 4269, but trying to stack them gives

```
ValueError: Cannot determine common CRS for concatenation inputs, got ['NAD83', 'NAD83']. Use `to_crs()` to transform geometries to the same CRS before merging.
```

even though i can verify `df1.crs.equals(df.crs)`. 

Data and details here https://gist.github.com/knaaptime/8e3b0841facd39856495dcb4923ffb16

**Comment by knaaptime** [comment]
adding to the mix, other files ([like 2017](https://www.dropbox.com/scl/fi/smg3kfk2r3k041s4nbhhn/dc17.parquet?rlkey=co4k7lpkscc8koxhk7pni2nua&dl=0)) play just fine (i.e. you can concat 17 and 18, just not 19) even though they're all 4269

**Comment by martinfleis** [comment]
@snowman2 do you have any idea why `==` considers two `pyproj.crs.CRS` objects equal but `set` does not?

**Comment by m-richards** [comment]
> A set hashes the CRS object and that uses the [`__hash__`](https://github.com/pyproj4/pyproj/blob/0c5159aab8c32488584ae472b3752203f1005559/pyproj/crs/crs.py#L1576-L1577) method to convert the CRS to WKT2 strings. The [`__eq__`](https://github.com/pyproj4/pyproj/blob/0c5159aab8c32488584ae472b3752203f1005559/pyproj/crs/crs.py#L1566-L1567) method uses the PROJ to compare whether the CRS are the eqivalent. Two CRS can be equivalent without having the same WKT2 string.

Is this considered correct behaviour in pyproj? Generally if objects are "equal" they should have the same hash  -  the [python docs](https://docs.python.org/3/reference/datamodel.html?highlight=__eq__#object.__hash__) say
>The only required property is that objects which compare equal have the same hash value; 

(I am not at all across the technical details of CRS, so perhaps this is the best compromise to have an efficient hash, and we certainly should be able to work around this on the geopandas side. But we might not be the only ones caught out by this)

**Comment by snowman2** [comment]
> Is this considered correct behaviour in pyproj? 

It is the best that we were able to come up with. The logic PROJ uses to compare CRS is complex and is difficult to be simplified to a hash. Ideas and contributions are welcome to improve this behavior 😃

**Comment by snowman2** [comment]
Related https://github.com/pyproj4/pyproj/issues/303

**Comment by m-richards** [comment]
Going to close this issue now, since we do check metadata in merge and concat and have done for some time. Any related discussions are probably better served in a new issue.

**Comment by knaaptime** [comment]
@m-richards I still have the same issue shown [this gist](https://gist.github.com/knaaptime/8e3b0841facd39856495dcb4923ffb16) on 14.2. Woud you prefer if i open another issue?

**Comment by knaaptime** [comment]
hm. since #3023 was merged in november and 14.2 was released 2 weeks ago, i figured 14.2 includes that patch?

**Comment by knaaptime** [comment]
ah gotcha. thanks, will keep an eye out
